aws-s3-private-bucket/main.tf

```
@@ -10,7 +12,31 @@ locals {
 
 resource "aws_s3_bucket" "bucket" {
   bucket = var.bucket_name
-  acl    = "private"
+  # To define the access control list (ACL) of the bucket, there are two ways:
+  # 1. Define the canned ACL using the acl argument
+  # 2. Define the grant ACL using the grant argument
+  # These two ways conflict with each other so we can only use one of them
+
+  # Using canned ACL will conflict with using grant ACL
+  acl = local.use_grant_acl ? null : "private"
+
+  # Use ACL policy grant to grant permissions to certain users, grant will conflict with canned ACL.
+  # We sometimes need to grant ceertain canonical users, 
+  # (eg. when it is a logging bucket, then we need to grant the awslogsdelivery account FULL_CONTROL permission
+  # more details of how to grant awslogsdelivery permissions in https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/AccessLogs.html)
+  # Details in how to find the canonical user id: https://docs.aws.amazon.com/general/latest/gr/acct-identifiers.html
+  dynamic "grant" {
+    for_each = [for pair in var.canonical_user_id_and_permissions : {
+      canonical_user_id = pair.canonical_user_id
+      permissions       = pair.permissions
+    }]
+
+    content {
+      id          = grant.value.canonical_user_id
+      permissions = grant.value.permissions
+      type        = "CanonicalUser"
+    }
```

I'm pretty sure this is valid, and if so it would be simpler: 

```suggestion
    for_each = var.canonical_user_id_and_permissions 

    content {
      id          = pair.canonical_user_id
      permissions =  pair.permissions
      type        = "CanonicalUser"
    }
```

but if I am wrong, nevermind.

---

looks like we still need the original way, otherwise it complains `canonical_user_id` not declared